### PR TITLE
A more robust Theme Manager

### PR DESCRIPTION
**Pull request by UTDZac**
This PR improves upon the recently added theme thumbnail preview. There are quite a few new things here:

- You can access the "**Theme Manager**" just as before by clicking Customize Theme -> More Options
- Each preloaded theme and user-added custom theme is properly rendered in the thumbnail preview
- Scroll through the themes seamlessly with arrow navigation keys, or look up from a list using the magnifying glass
- The `ThemePresets.txt` file is no longer part of the repository, but rather populated when the user first starts up their Tracker. Previously created preset files will still function as before
- This feature offers compatibility support for old theme presets, especially old theme codes. It should load them up fine, match a name to each theme, and display them properly

#### Button Functions
- Clicking "**Apply Theme**" will immediately apply the theme from the preview onto the entire Tracker
- If instead you are editing your Trackers theme, or perhaps imported a new one, this button becomes "**Save as new**" to create a new theme in your theme library
- A "**Remove**" button is added so that the user can neatly clean up any old themes they no longer want to keep around
- "**Import**" works same as before, but "**Export**" now exports the _currently previewed theme_, so you can share with your friends without first having to apply that theme to your own Tracker.

image

### CODE REVIEW

File: ironmon_tracker/Theme.lua [hunk 1]
```diff
@@ -361,38 +440,128 @@ end
...
+				table.insert(Theme.PresetsOrdered, themeName)
+			end
+
+			Utils.addCustomThemeToFile(themeName, themeCode)
+			Theme.PresetStrings[themeName] = themeCode
+			Theme.refreshThemePreview()
+
+			client.unpause()
+			forms.destroy(form)
+		end
```
[the remark this answers is not in the document]
Comment (UTDZac) [reply ironmon_tracker/Theme.lua]: Agree. I've now included a warning in red text stating what went wrong. There are currently three conditions where the save can fail:

1. The user enters a reserved theme name such as "Current Theme (Custom)" or "Default Theme"
2. The user enters a theme name that contains six consecutive hexcode characters (0-9A-F) which would end up breaking the parsing later
3. The user enters a name for an existing theme, they will first be prompted to confirm they wish to overwrite before allowing it.

File: ironmon_tracker/Theme.lua [hunk 2]
```diff
@@ -361,38 +440,128 @@ end
...
 end
 
+function Theme.openSaveCurrentThemeWindow()
+	Program.destroyActiveForm()
+	local form = forms.newform(290, 130, "Save Theme As...", function() client.unpause() end)
+	Program.activeFormId = form
+	Utils.setFormLocation(form, 100, 50)
+
+	forms.label(form, "Enter a name for this Theme:", 18, 10, 300, 20)
+	local saveTextBox = forms.textbox(form, "", 200, 30, nil, 20, 30)
```
[the remark this answers is not in the document]
Comment (UTDZac) [reply ironmon_tracker/Theme.lua]: Easy enough fix. I added in a max length limit, arbitrarily picked `80`.

File: ironmon_tracker/Theme.lua [hunk 3]
```diff
@@ -361,38 +440,128 @@ end
...
+		Theme.importThemeFromText(themeCode, false)
+
 		client.unpause()
 		forms.destroy(presetsForm)
 	end, 212, 29)
 end
 
+function Theme.openSaveCurrentThemeWindow()
+	Program.destroyActiveForm()
+	local form = forms.newform(290, 130, "Save Theme As...", function() client.unpause() end)
```
[the remark this answers is not in the document]
Comment (UTDZac) [reply ironmon_tracker/Theme.lua]: Whoops, somehow glanced over this one. Fixed it up, should look nicer now.

File: ironmon_tracker/Theme.lua [hunk 4]
```diff
@@ -361,38 +446,160 @@ end
...
+				forms.settext(Theme.Manager.SaveNewConfirm, "Save")
+			elseif Theme.PresetStrings[themeName] ~= nil then
+				-- If the Theme name is already in use, warn the user first
+				if forms.gettext(Theme.Manager.SaveNewConfirm) ~= "Confirm" then
+					forms.settext(Theme.Manager.SaveNewWarning, "A Theme with that name already exists. Overwrite?")
+					forms.settext(Theme.Manager.SaveNewConfirm, "Confirm")
+				else
+					success = true
+				end
+			end
```
Comment (Fellshadow) [inline ironmon_tracker/Theme.lua:535]: The code here only ever sets `success` to true when you press confirm on overwriting an old theme, meaning actual valid inputs of a theme name such as "Test" will not save with no warning messages or anything (the save button just does nothing)

File: ironmon_tracker/Theme.lua [hunk 5]
```diff
@@ -361,38 +446,160 @@ end
...
 
+function Theme.openSaveCurrentThemeWindow()
+	Program.destroyActiveForm()
+	local form = forms.newform(350, 145, "Save Theme As...", function() client.unpause() end)
+	Program.activeFormId = form
+	Utils.setFormLocation(form, 100, 50)
+
+	forms.label(form, "Enter a name for this Theme:", 18, 10, 330, 20)
+	local saveTextBox = forms.textbox(form, "", 290, 30, nil, 20, 30)
+	forms.setproperty(saveTextBox, "MaxLength", 80)
```
Comment (Fellshadow) [inline ironmon_tracker/Theme.lua:504]: 80 is a fine value, from a quick test seems somewhere around 28-ish seems like the limit that fits on the screen (depending on the characters used) if interested in limiting it by that. Disregard if not too concerned about that

**Review by Fellshadow — CHANGES_REQUESTED** [review]
Like the changes made from last one, though the logic for the save button is broken now 👀

File: ironmon_tracker/Theme.lua (hunk 4 above)
Comment (UTDZac) [reply ironmon_tracker/Theme.lua]: Good catch. I don't know why I wasn't just using `return` to escape out of the function when an issue arose. It should be fixed now.

**Comment by UTDZac** [comment]
While fiddling with these fixes, I finally caved and swapped the flow of the Theme Manager. Now the user will see the "Theme Library" first when clicking "Customize Theme" from the Navigation Menu screen. From here, they can edit the colors of the active theme. This approach I feel is a bit more intuitive, and lets the user explore other theme options right away.

image

**Review by Fellshadow — APPROVED** [review]
Yup, all good now and appears to work fine for me. Definitely prefer the new flow of the screens

**Pull request merged**